### PR TITLE
Added functionality to archive/unarchive teams from system console

### PR DESCRIPTION
**Pull request by manojmalik20**
#### Summary
Added feature to archive/unarchive teams from the system console

#### Screenshots

image

image

image

#### Release Note

```release-note
Added new feature to archive and unarchive teams through the Teams section in the system console
```

### CODE REVIEW

**Comment by mkraft** [comment]
@esethna I've added you for PM review and @matthewbirtch for UX review. Please redirect as appropriate.

**Review by matthewbirtch — CHANGES_REQUESTED** [review]
Thanks for the contribution @manojmalik20! Seems like a great idea to have this feature.

Just a couple notes from me:
1. I can't seem to consistently reproduce this, but when I archive a team and save that change, it kicked me out of the system console and back in to the chat UI. Can we make sure it always routes back to the Teams listing page in the system console?
2. If I archive the currently active team, and I click 'back to {team}' in the top-left, it errors out with 'Team not found'. This makes sense, but we should consider activating the last accessed unarchived team and changing the 'back to {team}' so that it reflects that team.
3. If the server only has one team that is not archived, should we allow archiving of the one team? I wonder if the archive button should be disabled or hidden in this case - maybe with some help text to explain that 'Archiving is disabled when there is only one active team`. I tried archiving the last remaining team and as soon as I archived the last team, it kicked me out of the system console to this screen below. 
image

4. For the teams list view, can we change the `(Archived)` color so that it's not red? Red implies an error. I would suggest going with a 56% or 64% opacity of the standard text color

File: components/admin_console/team_channel_settings/team/details/team_details.jsx
Comment (mickmister) [inline components/admin_console/team_channel_settings/team/details/team_details.jsx:125]: Can we have this instance of `saveNeeded` scoped to this block, since we return at the end of the block? This way there's no confusion on whether "changing" that value here could affect code below this block.
```suggestion
        if (this.teamToBeArchived()) {
            let saveNeeded = false; // add this line here
            const result = await actions.deleteTeam(team.id);
            if ('error' in result) {
                serverError = <FormError error={result.error.message}/>;
                saveNeeded = true;
```

File: components/admin_console/team_channel_settings/team/details/team_details.jsx
Comment (mickmister) [inline components/admin_console/team_channel_settings/team/details/team_details.jsx:118]: This line would then go below the `if (this.teamToBeArchived())` and `else if (this.teamToBeRestored() && this.state.serverError === null)` blocks, to be a separate variable declaration.

File: components/admin_console/team_channel_settings/team/details/team_details.jsx
Comment (mickmister) [inline components/admin_console/team_channel_settings/team/details/team_details.jsx:505]: It seems odd that the save button and modal title both say "Save and Archive Team". Maybe the button should just say "Archive"?

File: components/admin_console/team_channel_settings/team/details/team_details.jsx
Comment (mickmister) [inline components/admin_console/team_channel_settings/team/details/team_details.jsx:410]: I'm not sure that I understand this logic. Why is toggling the archive button showing and hiding a server error? Is it possible that the error has nothing to do with archiving? I'm thinking we should show the error until we have another form submission.

**Review by mickmister — CHANGES_REQUESTED** [review]
Thanks for the PR @manojmalik20! Mostly looks good :+1:  I have a few requests

**Comment by esethna** [comment]
Adding @wiersgallak @aaronrothschild as they might have more context on system console features. If the concept of "archiving a team" didn't exist before this, then I think we need to think through what that actually means, and what happens to existing channels, users, memberships, accounts etc on that team.

File: components/admin_console/team_channel_settings/team/details/team_details.jsx
Comment (manojmalik20) [reply components/admin_console/team_channel_settings/team/details/team_details.jsx:410]: Suppose there is a server error that is because of removing or adding a user in the team, but now the user is archiving the team so those server errors are irrelevant now and if the user has just toggled and untoggled the archive button then those errors should persist in the state. Also, on the channel details screen, the server errors are handled in the same way on archiving/unarchiving the channel. That's why I added the same logic here.

File: components/admin_console/team_channel_settings/team/list/team_list.tsx
Comment (mickmister) [inline components/admin_console/team_channel_settings/team/list/team_list.tsx:232]: Can we add parens here as well?

```suggestion
                                    {team.delete_at !== 0 && (
                                        <span className='archived-label'>
```

**Review by mickmister — COMMENTED** [review]
> I was wondering how to stop that event handler to stop redirecting if the event is fired because of the `Archive Team` option in the system console.

Since the system console is separate from the channel view, can we decouple their relationship as far as the "current team"?  We could have conditional logic in the websocket handler for handling the case of being in the system console, that effectively sets the currently active team to null. Then when the user tries to navigate back to the channel view, that's when we send them to the select team page. @manojmalik20 What are your thoughts on this?

**Comment by aaronrothschild** [comment]
> Adding @wiersgallak @aaronrothschild as they might have more context on system console features. If the concept of "archiving a team" didn't exist before this, then I think we need to think through what that actually means, and what happens to existing channels, users, memberships, accounts etc on that team.

I agree with this....what happens to channels?  How can I find them again?  Unarachive the team then unarchive the channels?  

@manojmalik20 Right now, does this essentially "hide" a team without deleting it?

**Comment by manojmalik20** [comment]
> I agree with this....what happens to channels? How can I find them again? Unarchive the team then unarchive the channels?
> 
> @manojmalik20 Right now, does this essentially "hide" a team without deleting it?

Yes @aaronrothschild, this essentially "hides" a team without deleting it. Once you unarchive the team, you don't have to separately unarchive channels unless they were already archived before archiving the team. So basically, on unarchiving a team, it restores back to the state in which it was before archiving.

**Comment by manojmalik20** [comment]
> We could have conditional logic in the websocket handler for handling the case of being in the system console, that effectively sets the currently active team to null. Then when the user tries to navigate back to the channel view, that's when we send them to the select team page. @manojmalik20 What are your thoughts on this?

@mickmister Yes, I agree with you. Sending the user to the select team page instead of finding the previously active unarchived team seems reasonable. So, I'll make that change.

**Comment by manojmalik20** [comment]
@mickmister Yes, I agree with you. But instead of sending the user to select team page directly, what I've done is I'm sending the user to the homepage (eg - `http://{mattermost-url}/`) and if the user has a previously active unarchived team then the user gets redirected to that team and if he doesn't have such a team, then he gets redirected to the select team page. What do you think about this?

**Comment by manojmalik20** [comment]
@matthewbirtch Can you please re-review this?

**Comment by matthewbirtch** [comment]
@manojmalik20 it looks like the issues that I identified are resolved,

>  instead of sending the user to select team page directly, what I've done is I'm sending the user to the homepage (eg - http://{mattermost-url}/) and if the user has a previously active unarchived team then the user gets redirected to that team and if he doesn't have such a team, then he gets redirected to the select team page. What do you think about this?

I think this could work, although it doesn't feel overly smooth with the redirects and flashes of other screens in that process. Not sure if there is any way to avoid that flicker?

**Comment by manojmalik20** [comment]
@matthewbirtch That flash of other screen is because we are directing the user to the "/" path. That flash can be removed by redirecting the user directly to the `select_team` screen, but we cannot do that because if the user is part of an unarchived team, he should be redirected to that team instead of the `select_team` page. So I guess we cannot remove that flicker. If there's any other solution, maybe the developers can help me out. @mickmister

**Review by matthewbirtch — APPROVED** [review]

**Comment by manojmalik20** [comment]
This PR is ready for re-review @mickmister @mkraft @aaronrothschild @wiersgallak

**Review by mkraft — APPROVED** [review]

**Comment by esethna** [comment]
@manojmalik20 really appreciate the PR. 

I have some thoughts, I'm not sure if some of them were addressed since there's no ticket for this. I'm worried there's a lot to the idea of "Archiving a Team" that we may not be considering (correct me if I'm wrong), off the top of my head here are a few: 

1. What happens to users where this was their only team? If they get dropped at the select team page how are users informed what happened or where their team went?
2. What happens to users where this was one of a few teams they were members of? Does it hide the team in the team sidebar? If so, how is the user informed of why they can't find that team anymore?
3. What about the above cases if the user is logged in via mobile?
4. What if a user hits a permalink or email notification to the archived team?
5. What if the user taps a push notification to the archived team?

I'm sure there's more considerations than the above. I think this is a large feature and probably shouldn't be taken without a ticket and expected behaviour that's agreed on by UX and PM. My proposal would be to close this PR until that exercise is completed. 

cc// @matthewbirtch @aaronrothschild

**Comment by chetanyakan** [comment]
@esethna Before we decide to close this PR, I just wanted to point out that the functionality to archive/unarchive teams already exists in Mattemost CLI commands.
This PR just exposes the functionality to the system console UI.

Maybe it makes sense to brand this feature as `beta`/`experimental` with some warnings instead of closing the PR entirely. (1/5)

**Comment by esethna** [comment]
Hmm thanks @chetanyakan I wasn't aware. The challenge there is that but surfacing the UI button to archive, we are promoting the use of this feature (even if our docs label it as experimental or beta). And without the comprehensive design I'm worried about admins using this feature with the expectation that all of the above considerations are handled gracefully.

**Comment by esethna** [comment]
@chetanyakan Can you help me understand the use cases for needing this feature? Maybe that will help with understanding how important the considerations I listed above might be?

**Comment by esethna** [comment]
@mkraft @sbishel any thoughts on [the above ](https://github.com/mattermost/mattermost-webapp/pull/8129#issuecomment-863495151), maybe you guys are more familiar with the use cases for this feature and if these concerns are warranted given the ability to archive a team already exists in the CLI

**Comment by mkdbns** [comment]
@esethna , the use case for this is pretty simple. There are customers who have MM System Admins that want to be able to archive / unarchive teams but do not have access to the host / cli - or they don't want to go through the hassle of ssh-ing in. This just gives them the ability to do so via the system console just like they can do for channels. Honestly, it seems like a gap in the Teams Admin UI that we're just trying to improve on. Does that make sense?

**Comment by esethna** [comment]
Thanks @mkdbns 

@wiersgallak looping you in here as I think you might be more familiar with this area of the product than I am. Given that archiving a team already exists via the CLI I'm not opposed to accepting this PR. 

With that said, regardless of if an admin is archiving a team from the CLI or these new system console buttons, we should recognize there are likely gaps in the UX of this feature from the end-user perspective as I noted above. However, if the typical use case for archiving a team is one where end users are no longer active in the team, and they wouldn't notice if it simply disappeared, then maybe my concerns are not warranted (and could potentially be solved by documentation).  

@wiersgallak I will leave it your hands to approve this PR or not, given your understanding of typical admin workflows and any data you might have on requests for this feature.

**Review by wiersgallak — APPROVED** [review]
Thanks, Eric.  This is approved from my perspective.  There are some edge cases that we will need to build upon, but we can queue some additional tickets for this.  This definitely adds value in that the action is now exposed in the UI.  We will ensure that we add the nuances of the ideal situation for end-users in the documentation of this feature until we are able to iterate on the experience.

File: components/admin_console/team_channel_settings/team/details/team_profile.jsx
Comment (mickmister) [inline components/admin_console/team_channel_settings/team/details/team_profile.jsx:72]: I personally think the comments regarding class names are not needed here. We usually want to include comments when there is logic that needs to be explained. I think we are fine without the comment here and the one below.

Though if we do include the comments, they should use the same comment style:
```suggestion
                    // We are using same classes used for Archive Channel button so that we can use the same styling as Archive channel button
```

**Review by mickmister — APPROVED** [review]
Thanks for implementing the changes @manojmalik20 :+1: I'm approving with one non-blocking request regarding the code comments added in the last commit.

**Comment by manojmalik20** [comment]
@mickmister I've removed the comments.

**Comment by mickmister** [comment]
@manojmalik20 Please avoid force-pushing when possible. When a PR's commits are force-pushed, the commit history of the PR is not maintained, so there's no way to tell what has changed since last review, and so the PR needs to be reviewed again. Especially with an open-source project, this is very important as we need to be certain about any code that is introduced.

Hopefully this makes sense 🙂

**Comment by manojmalik20** [comment]
@mickmister Yeah this makes sense. I'll keep this in mind from now on.

**Comment by manojmalik20** [comment]
@mkraft @mickmister @wiersgallak Is there anything left on this PR?

**Comment by jasonblais** [comment]
Sorry for the delay, and thanks for the contribution @manojmalik20! 

The last step is a QA review, I've requested @furqanmlk to help. Once completed, we're ready to merge! :)

**Review by furqanmlk — COMMENTED** [review]
N/A

**Review by furqanmlk — APPROVED** [review]
Working as expected. all e2e test cases passed ✅

**Comment by jasonblais** [comment]
Thanks again @manojmalik20! 

cc @amyblais not sure if this should have a Jira ticket for release test coverage.

**Pull request merged**
